libcephfs,client: zero copy interface for preadv/pwritev

salieri11 opened the pull request:
This PR replaces the [original](https://github.com/ceph/ceph/pull/58228) PR by @ffilz .

This pull request is in support of
https://jsw.ibm.com/browse/ISCE-754

Tracker
https://tracker.ceph.com/issues/69919 

NFS: Enable Async IO and Zero Copy in Ganesha FSAL Plugin
**Note:** zerocopy and object cacher are mutually exclusive at this point, in order to use the new API with zerocopy enabled one must disabled the object cacher.
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

vshankar commented:
@salieri11 - fs suite has tests that run fit with nfs-ganesha with OC disabled. See `qa/suites/fs/nfs/tasks/nfs.yaml`. The test itself is in `qa/tasks/cephfs/test_nfs.py::test_async_io_fio()`.

ffilz commented on src/libcephfs.cc:
```diff
@@ -2127,13 +2127,13 @@ extern "C" void ceph_ll_readv_writev_release(void *release_data)
   ceph_ll_readv_writev_buffer *buf = new ceph_ll_readv_writev_buffer;
 
   if (!io_info->write && io_info->zerocopy) {
-    ceph_assert(io_info->iovcnt == 1);
```
We need this test for io_info->iovcnt in the if since we are no longer doing it in assert

So the if test should be:

// a zero copy read MUST provide a length 1 iovec (where the buffer is a nullptr and the length is the requested read length
if (!io_info->write && io_info->zerocopy && (io_info->iovcnt != 1) {

salieri11 replied:
i dont know how i missed it, we talked about it on slack. fixing.

ffilz reviewed (commented):
A couple comments

salieri11 commented:
Thanks, this is draft PR, so I will do all the changes and then convert it to reviewable PR. But I guess since you are already familiar with the changes in the original PR, it will be easier to review this one.

First teuthology run, for sanity, only `test_nfs` test from the `fs` suite, as suggested by @vshankar
https://pulpito.ceph.com/igolikov-2025-02-26_17:39:55-fs-wip-igolikov-frank-zcopy-testing-default-smithi/

salieri11 commented on src/client/Client.cc:
```diff
@@ -11371,6 +11371,8 @@ int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
     } else {
         bufferlist bl;
+        // if zerocopy, blp can't be null due to the check above
+        // so the read will be into the memory provided by caller
```
oh right, just got fuzzy.

ffilz commented on src/client/Client.cc:
```diff
@@ -11345,6 +11345,10 @@ int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
+    // negative iovcount is invalid,
+    // if we are doing zerocopy, we must also have a bufferlist
+    // zero copy could work with a passed in buffer for write 
+    // (but not for read - for read, we need an iovec to fill, a passed in buffer would force a copy).
```
Hmm, though here, the param is a bufferlist... We COULD take a bufferlist that we fill...

salieri11 replied:
So maybe the correct check will be

```
iif (iovcnt < 0 || (zerocopy && (!blp || (!write && !iov)))) {
    return -EINVAL;
}
```
And it can be used both in `libcephfs.cc` and `Client.ccc` ?

ffilz commented on src/libcephfs.cc:
```diff
@@ -2107,6 +2107,10 @@ extern "C" void LL_onfinish_release(void *release_data)
 extern "C" int64_t ceph_ll_nonblocking_readv_writev(class ceph_mount_info *cmount,
 						    struct ceph_ll_io_info *io_info)
 {
+  if (!io_info->write && io_info->zerocopy && io_info->iovcnt != 1) {
```
Should add a comment here... Same comment as added to ceph_ll_readv_writev_buffer below

salieri11 commented:
> Thanks, this is draft PR, so I will do all the changes and then convert it to reviewable PR. But I guess since you are already familiar with the changes in the original PR, it will be easier to review this one.
> 
> First teuthology run, for sanity, only `test_nfs` test from the `fs` suite, as suggested by @vshankar https://pulpito.ceph.com/igolikov-2025-02-26_17:39:55-fs-wip-igolikov-frank-zcopy-testing-default-smithi/

This run fails, will rerun with `--debug=all` for the `fio` tool

dparmar18 commented:
> > Thanks, this is draft PR, so I will do all the changes and then convert it to reviewable PR. But I guess since you are already familiar with the changes in the original PR, it will be easier to review this one.
> > First teuthology run, for sanity, only `test_nfs` test from the `fs` suite, as suggested by @vshankar https://pulpito.ceph.com/igolikov-2025-02-26_17:39:55-fs-wip-igolikov-frank-zcopy-testing-default-smithi/
> 
> This run fails, will rerun with `--debug=all` for the `fio` tool

This failure is being tracked [here](https://tracker.ceph.com/issues/69347), it fails at doing random async read, so far the issue doesn't seem to be CephFS, **it's worth noting this only fails with NFS mounts with V4.2 on smithi nodes, V4.2 + local/qe machines work fine**. If you use older versions, they work fine too. I'm not sure what's causing the fio command to fail, first it was suspected to be SELinux (from kernel ring buffer) but couldn't find relevant or any solid evidence apart from selinux acls gettattr being returned -61 but i guess its not that of an issue. https://github.com/ceph/ceph/pull/61547 enables libcephfs logging for nfs clients where I could not find any trace of libcephfs layer doing anything. From my recent conversation with venky, as a stop gap, we will make use of 4.1 instead of 4.2 to at least move ahead with all the nfs developments and revert back once we push have the RCA. 

Tl;dr - It looks like the way smithi nodes are setup, isn't gelling well with NFS V4.2

dparmar18 commented:
Can you try running with this patch https://github.com/ceph/ceph/pull/62031 @salieri11 ?

salieri11 commented:
> Can you try running with this patch #62031 @salieri11 ?

Sure, will try it right now and update later

salieri11 commented:
> Can you try running with this patch #62031 @salieri11 ?

https://pulpito.ceph.com/igolikov-2025-02-27_14:06:55-fs-wip-igolikov-frank-zcopy-testing-default-smithi/
Green wiuth the patch

salieri11 commented:
Added sanity unit test for `libcephfs` functions being called with ZC flag, to verify that invalid combination of parameters will fail.
Once we will be OK on the content of this PR, I will rearrange all commits and prepare it for final review.

salieri11 commented:
After rebasing and adding blocking write test. WIll add blocking read and non blocking r/w tests to complete the coverage

salieri11 commented:
jenkins please retest this

salieri11 commented:
jenkins test make check

salieri11 commented:
jenkins test make check arm64

ffilz committed:
cephfs client: add zero-copy capabilit

For writes, the incoming iovec buffers are pushed into the bufferlist.

For reads, an incoming iovec is used to communicate the desired read
length, but then on read complete, an iovec would be built from the
bufferlist and passed back to the client.

Fixes: https://tracker.ceph.com/issues/69919
Signed-off-by: Igor Golikov <igolikov@ibm.com>

ffilz committed:
libcephfs: expose zerocopy in libcephfs

Zero copy can be used both with and without non-blocking I/O.

Signed-off-by: Igor Golikov <igolikov@ibm.com>
Fixes: https://tracker.ceph.com/issues/69919

salieri11 committed:
cephfs client: add RW copy metrics

add metrics for counting copy ops and copy sizes
these metrics are used by zerocopy tests

Signed-off-by: Igor Golikov <igolikov@ibm.com>
Fixes: https://tracker.ceph.com/issues/69919

salieri11 committed:
test: add sanity and integration test for zerocopy

add sanity unit test for libcephfs zerocopy illegal params
add integration test as a part of the Client test for zerocopy RW

Signed-off-by: Igor Golikov <igolikov@ibm.com>
Fixes: https://tracker.ceph.com/issues/69919

ffilz commented on src/libcephfs.cc:
```diff
@@ -2080,31 +2080,157 @@ extern "C" int64_t ceph_ll_writev(class ceph_mount_info *cmount,
+
+  io_info->result = (cmount->get_client()->ll_preadv_pwritev(
+			io_info->fh, io_info->iov, io_info->iovcnt,
+			io_info->off, io_info->write, nullptr, &buf->bl,
```
I think we can pass nullptr instead of &buf.bl for writes

salieri11 replied:
agree, i don't see any use of bufferlist for writes.

ffilz commented on src/libcephfs.cc:
```diff
@@ -2080,31 +2080,157 @@ extern "C" int64_t ceph_ll_writev(class ceph_mount_info *cmount,
+extern "C" int64_t ceph_ll_readv_writev(class ceph_mount_info *cmount,
+				     struct ceph_ll_io_info *io_info)
+{
+  ceph_ll_readv_writev_buffer *buf = new ceph_ll_readv_writev_buffer;
```
I think we only need buf for reads... We should verify all the assumptions and not allocate for writes if we don't need

ffilz commented on src/client/Client.cc:
```diff
@@ -11343,23 +11351,39 @@ int Client::pwritev(int fd, const struct iovec *iov, int iovcnt, int64_t offset)
-    if(iovcnt < 0) {
+    // negative iovcount is invalid,
+    // for ZC read - we must use provided buffer, it will contain the data after the read
+    // for ZC write - we must have iovec AND we will enforce that CEPH buffer is not passed
```
No, look at ceph_ll_preadv_pwritev and ceph_ll_nonblocking_readv_writev

We always pass a bufferlist, and if read we also expect a length 1 iovec (actually we don't make the read check in ceph_ll_nonblocking_readv_writev...)

For write, the bufferlist isn't actually used... Maybe we should change the whole stack to pass nullptr for bufferlist for writes... Need to check all the assumptions there...

salieri11 replied:
I understand the write part. Cleaning up the whole stack can be done as a separate task.
As for the reads, the copy from bufferlist to iovec is done only if not ZC:
```
if (write) {
        int64_t w = _write(fh, offset, totallen, NULL, iov, iovcnt, onfinish,
                           do_fsync, syncdataonly, zerocopy);
        ldout(cct, 3) << "pwritev(" << fh << ", \"...\", " << totallen << ", " << offset << ") = " << w << dendl;
        return w;
    } else {
        bufferlist bl;
        // if zerocopy, blp can't be null due to the check above
        // the read will be into Ceph buffers which are returned to the caller.
        int64_t r = _read(fh, offset, totallen, blp ? blp : &bl,
                          onfinish);
        ldout(cct, 3) << "preadv(" << fh << ", " <<  offset << ") = " << r << dendl;
        if (r <= 0) {
          return r;
        }

	if (!zerocopy) {
          client_lock.unlock();
          uint64_t size  = copy_bufferlist_to_iovec(iov, iovcnt, blp ? blp : &bl, r);
          // all IOVs will be copied, the function above returns the total bytes being copied
	        update_total_copy_size(size, iovcnt);
          client_lock.lock();
        }
        return r;
    }
```
And for ZC reads nothing is copied, the read is into the bufferlist.
Since the writes are not using bufferlist at all, I commented the check very explicitly: 
for writes, we must have iovec, and we don;t care about the bufferlist, and for reads - vice versa
Am I missing something here?

ffilz commented on src/client/Client.cc:
```diff
@@ -11343,23 +11351,39 @@ int Client::pwritev(int fd, const struct iovec *iov, int iovcnt, int64_t offset)
     }
     loff_t totallen = 0;
     for (int i = 0; i < iovcnt; i++) {
+        // NOTE: zero copy read has passed in an iov that is just used to
```
passed in an iov of length 1 that is just used

vshankar commented:
I will run this through relevant test as an exercise to know where we stand in terms of breakages (if any).

vshankar commented:
This PR is under test in [https://tracker.ceph.com/issues/70647](https://tracker.ceph.com/issues/70647).

salieri11 committed:
libcephfs,client: fix parameters validation for ZC

Signed-off-by: Igor Golikov <igolikov@ibm.com>
Fixes: https://tracker.ceph.com/issues/69919

salieri11 committed:
test: do not install unit test binary

as it follows from the codebase, there is no need to install unit test
binaries, only the integration test binaries are installed during build

Signed-off-by: Igor Golikov <igolikov@ibm.com>
Fixes: https://tracker.ceph.com/issues/69919

xhernandez commented on src/include/cephfs/libcephfs.h:
```diff
@@ -120,15 +120,19 @@ struct ceph_snapdiff_entry_t {
 
 struct ceph_ll_io_info {
   void (*callback) (struct ceph_ll_io_info *cb_info);
+  void (*release) (void *);
+  void *release_data;
   void *priv; // private for caller
   struct Fh *fh;
-  const struct iovec *iov;
+  struct iovec *iov;
   int iovcnt;
+  int iovmax; // maximum iovcnt is allowed, 0 for no limit
   int64_t off;
   int64_t result;
   bool write;
   bool fsync;
   bool syncdataonly;
+  bool zerocopy;
 };
```
I think the modification of `struct ceph_ll_io_info` is dangerous. It makes the structure binary incompatible with older versions, which means that any existing program using async I/O from the previous libcephfs.so will break once the new version is installed.

Even moving the fields at the end will be problematic because the applications using the previous version won't initialize them and will send garbage, which causes failures.

salieri11 replied:
@ffilz WDYT? Any issues with binary incompatibilty?

vshankar replied:
I had a conversation with @xhernandez last week and this is an issue for _independent_ applications that are linked to libcephfs and there is no guarantee that upgrading the library would need to recompile the application. Since the application would use the structure from the header which on library upgraded introduced these new fields, many assumptions in the application regarding the header would break. So, we need a mechanism to have the application and the library be compatible by either exchanging feature sets on initialisation or have the library tell the application about what's supported in the structure.

kotreshhr replied:
I agree with @kalebskeithley here

kalebskeithley replied:
I'd venture to say that you should bump LIBCEPH_VER_{MAJOR,MINOR,EXTRA} in libcephfs.h whenever you make an ABI change.

Have there really been no ABI changes since 2021?

vshankar replied:
> I'd venture to say that you should bump LIBCEPH_VER_{MAJOR,MINOR,EXTRA} in libcephfs.h whenever you make an ABI change.

Valid point. 

> 
> Have there really been no ABI changes since 2021?

There have been API additions and I think Ganesha has always ensured the correct library version is used (e.g., nonblocking APIs). But, in general, an API change is also a candidate for bumping up the library version like it was done when introducing ceph_ll_register_callbacks2() in 2021...

anoopcs9 replied:
> I'd venture to say that you should bump LIBCEPH_VER_{MAJOR,MINOR,EXTRA} in libcephfs.h whenever you make an ABI change.

Irrespective of what those defines are we always build the library as _libcephfs.so.2.0.0_ 🤷🏻‍♂️ . As per [src/CMakeLists.txt](https://github.com/ceph/ceph/blob/main/src/CMakeLists.txt#L855),
```
    set(libcephfs_version 2.0.0)
    set(libcephfs_soversion 2)
    set_target_properties(cephfs PROPERTIES
      OUTPUT_NAME cephfs
      VERSION ${libcephfs_version}
      SOVERSION ${libcephfs_soversion})
```
we don't consume `LIBCEPHFS_VER_MAJOR`, `LIBCEPHFS_VER_MINOR` and `LIBCEPHFS_VER_EXTRA`.

vshankar replied:
Almost every other ceph library seems to be sticking to 2.0.0. See: src/librados/CMakeLists.txt

xhernandez replied:
> > Even moving the fields at the end will be problematic because the applications using
> > the previous version won't initialize them and will send garbage, which causes failures.
> 
> That's true, but it's probably more likely that crashes will occur in the library and not in the application that consumes the library.

What will crash is the entire process, so it's little relevant where the crash happens. However, what I've observed in this case is that we get unexpected failures from libcephfs, mostly with EINVAL error, caused by the garbage contents.

> The right way to do this is to add a new method using the new struct and preserve the old API/ABI until such time as _all_ the consumers (e.g. ganesha, samba) have had a chance to convert to the new API.

I agree. That would be a good backward compatible way of adding new functionality. The only thing to consider in this case is that if the application is upgraded before the library, some required symbols would be missing, so the application wouldn't start. That's harder to happen because normal package managers would make sure to also upgrade the library in this case, but manual builds could still have this issue.

vshankar replied:
OK, so it sounds like the way forward is to introduce a new method accepting the new structure. Ganesha would also require a change at this point to use the new methods. @ffilz ?

kalebskeithley replied:
Anoop wrote:
> Irrespective of what those defines are we always build the library as libcephfs.so.2.0.0 🤷🏻‍♂️ . As per 
> [src/CMakeLists.txt](https://github.com/ceph/ceph/blob/main/src/CMakeLists.txt#L855),
> ...
vshankar wrote:
> Almost every other ceph library seems to be sticking to 2.0.0. See: src/librados/CMakeLists.txt

this is the library's SO_NAME, and are not related, per se, to the LIBCEPHFS_VER_* defines.

The SO_NAMEs aren't changed across versions in order to allow seamless upgrades.

We could also use symbol versioning to preserve the ABI, as was done with libgfs in gluster, and while it's a subjectively better solution, it's also more complex.

If you're going to keep the same SO_NAME and not use versioned symbols then you need to take extra care not to break the ABI.

xhernanez wrote:
> What will crash is the entire process, so it's little [ir]relevant where the crash happens. 

Yes, we all know a SEGV anywhere crashes the entire process. It's not irrelevant where it happens. Crashing in the library that actually broke the ABI is far better and easier to debug than crashing in seemingly random places in the apps that consume the library and aren't aware that the ABI changed. I've only been writing software for 43+ years, so what do I know? A little I guess. Ganesha spent a week debugging this only to finally discover the bug is in libcephfs. <shrug>

gregsfortytwo replied:
Talked to @salieri11  today, and some comments (though I know this is re-treading a bit): we really should be bumping the LIBCEPHFS_VERSION fields whenever we make any changes (even compatible ones!), though we are very bad at it
> #define LIBCEPHFS_VERSION_CODE LIBCEPHFS_VERSION(LIBCEPHFS_VER_MAJOR, LIBCEPHFS_VER_MINOR, LIBCEPHFS_VER_EXTRA)

Even if we're not bumping the SO_NAME, users can still query the header file at build time and use that to make choices about what calls they're invoking.
My recollection is that we intended to increment MAJOR if we do an incompatible change, and otherwise bump the MINOR or EXTRA as we add functionality, depending on how big of an increment it is. (Does that sound fuzzy to you? It has always been fuzzy to me!)

As for ABI compatibility, yes, we for sure need to maintain that. This was a big oopsie! @salieri11 is working to resolve this

xhernandez replied:
I don't think that checking `LIBCEPHFS_VER*` at compile time is enough. The actual library installed in the system may not use the same version. Maybe it would be good to have a function similar to `ceph_version()` but returning the LIBCEPHFS_VER* values from libcephfs build time.

Another concern is that adding new functions to libcephfs is problematic because if the client uses them, they will be required at execution time. If the library is older, the execution will fail even before the application can check the current version and fall back to the previous implementation.

mattbenjamin replied:
it *is* enough for downstream products that pair up a ceph and ganesha of known versions, from build time to deployment.  That is the contract that was already in effect, which cephfs stopped honoring.

mattbenjamin replied:
All version bumps allow the library consumer to distinguish its need for...whatever is marked.  There should be no change that does not increment at least EXTRA.

salieri11 replied:
> LIBCEPH_VER_{MAJOR,MINOR,EXTRA} in libcephfs.h whenever you make an ABI change.

Since the latest changes preserve old API with full ABI compatability, please suggest which versions to bump.
In general, if we will preserve the approach I have taken with API versioning (e.g. next time the API will not change but only the length of the argument), I think that this time we DON'T need to bump the major version (we added new function) but we need to bump the minor one. @gregsfortytwo @vshankar WDYT?

ffilz commented:
Hmm, I hadn't considered mismatch of Ganesha and library versions...

Current Ganesha will COMPILE and RUN with any recent ceph library version, but a mismatch between what it's compiled against and what it runs against is tricky.

hcbwiz commented:
nfs-ganesha releases memory in the callback function:
   https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/FSAL/FSAL_CEPH/handle.c#L1845

Tt causes invalid accessing of **io_info** in  class LL_Onfinish::complete()

The lifecycle of **io_info** is controlled in the caller. So storing some flags in the constructer:
```
diff --git a/src/libcephfs.cc b/src/libcephfs.cc
index c6ff5f364..849412bd9 100644
--- a/src/libcephfs.cc
+++ b/src/libcephfs.cc
@@ -2108,14 +2108,16 @@ extern "C"  void LL_onfinish_release(void *);
   struct ceph_ll_readv_writev_buffer buf;
 private:
   // no ownership on io_info. should be released by the caller
   struct ceph_ll_io_info *io_info;
+  bool is_write;
+  bool zerocopy;
   void complete(int r) override {
     finish(r); // fills in io_info->result which may be different than r
-    if (io_info->result < 0 || io_info->write || !io_info->zerocopy) {
+    if (r < 0 || is_write || !zerocopy) {
       // this is an error, write, or a non-zerocopy read, delete...
       // for a succesful zero-copy read, we need to keep this object until the
       // caller is done with the buffers so it can use this as a hook to release   
```

salieri11 commented:
This PR will be refactored to supprot v2 struct

mattbenjamin commented:
> Hmm, I hadn't considered mismatch of Ganesha and library versions...
> 
> Current Ganesha will COMPILE and RUN with any recent ceph library version, but a mismatch between what it's compiled against and what it runs against is tricky.

I feel like this is solving a problem that downstream, at least, does not have.

vshankar commented:
> Current Ganesha will COMPILE and RUN with any recent ceph library version, but a mismatch between what it's compiled against and what it runs against is tricky.

This needs to be a part of our fs qa suite to catch such ABI breakages. Currently, we have tests that run `fio` with nfs-ganesha (introduced in https://github.com/ceph/ceph/pull/61143). @salieri11

mattbenjamin commented:
> > Current Ganesha will COMPILE and RUN with any recent ceph library version, but a mismatch between what it's compiled against and what it runs against is tricky.
> 
> This needs to be a part of our fs qa suite to catch such ABI breakages. Currently, we have tests that run `fio` with nfs-ganesha (introduced in #61143). @salieri11

catch it how?  are you introducing symbol/abi versioning?

vshankar commented:
> > > Current Ganesha will COMPILE and RUN with any recent ceph library version, but a mismatch between what it's compiled against and what it runs against is tricky.
> > 
> > 
> > This needs to be a part of our fs qa suite to catch such ABI breakages. Currently, we have tests that run `fio` with nfs-ganesha (introduced in #61143). @salieri11
> 
> catch it how? are you introducing symbol/abi versioning?

No, not at the moment.

By catch, I meant it would help in identifying such breakages if the Ganesha daemon crashes if we do things like running Ganesha with mismatching compiled vs runtime version.

salieri11 committed:
libcephfs: introduce v2 API for preadv/pwritev

The new API will ensure ABI backward compatibility, and introduce
explicit _v2 API for zerocopy supported calls.
The struct ceph_ll_io_info_v2 will be used to contain zero copy related
data.
The ceph_ll_io_info_wrapper struct is used as a helper struct to make
the usage of v1 and v2 structs easier.

Signed-off-by: Igor Golikov <igolikov@ibm.com>
Fixes: https://tracker.ceph.com/issues/69919

salieri11 commented:
Pushed new bits with versioned API and io_info struct length.
Need to fix tests, but the code can be reviewed (I would like to save time and to make a review while its WIP).
Since testing part will not change much and all the attention goes to the ABI compatability, please review.

xhernandez reviewed (commented):
From the backward compatibility point of view, I think this approach is good enough IMO.

salieri11 commented:
jenkins test make check arm64

xhernandez commented on src/include/cephfs/libcephfs.h:
```diff
@@ -119,16 +119,24 @@ struct ceph_snapdiff_entry_t {
+	void (*callback) (struct ceph_ll_io_info *cb_info);
+	void *priv; // private for caller
+	struct Fh *fh;
+	struct iovec *iov;
```
Is it really necessary to remove the `const` qualifier ? I don't think it will be critical, but it may cause issues in code compiled when the `const` qualifier was present. The compiler optimizations could be incorrect if the iovec is really modified.

salieri11 replied:
Well this is original change by @ffilz, who authored this part of the code while implementing zero copy (removing the `const`).
When handling non zerocopy request, the result is copied to `iov`:
https://github.com/ceph/ceph/blob/e9c5000a49b6b4a3eff273c2cdfbaf8af65a8d87/src/libcephfs.cc#L2415

xhernandez replied:
The contents of the buffers are not the problem. The problem is modifying the iovec vectors themselves. It seems they can be modified for reads with zerocopy enabled. If that's the only case, then I guess it's fine since for old clients the vectors will remain constant even if not explicitly declared. If that's the only case, I would say we can live with this change.

From the ABI compatibility point of view, I would say that this change is fine. I haven't reviewed the datails of the implementation of the zerocopy itself, though. I can take a look, but I'm not sufficiently familiarized with some parts of the code to do a proper review.

xhernandez replied:
Apparently I can't dismiss my own review requesting changes. If someone can, please dismiss it. The changes I requested are already addressed now.

salieri11 commented:
major library version has been set to 11

xhernandez commented on src/include/cephfs/libcephfs.h:
```diff
@@ -39,9 +39,9 @@ using CephContext = ceph::common::CephContext;
 #endif
 
-#define LIBCEPHFS_VER_MAJOR 10
+#define LIBCEPHFS_VER_MAJOR 11
```
If the changes only add new interfaces, shouldn't it be better to increase just the minor ?

salieri11 replied:
This what we discussed with the team and decided to bump major since it was not changed for the last 10 years. We want to make it now a baseline for the future changes.

gregsfortytwo replied:
hmm I also thought we were only bumping the major version on incompatible changes. @vshankar did you have a different opinion on this?

gregsfortytwo replied:
The argument being that you can check the major version to know very clearly if your code (compiled at some arbitrary point) will still run against the library, or if it's become incompatible.

salieri11 replied:
@vshankar WDYS?

vshankar replied:
The rules are pretty loose here. I suggested @salieri11 to bump the major version actually. But now I recall, the interfaces are new (the old implementation changed the ll_preadv_writev interface. So, I'm fine with bumping the minor version.

salieri11 commented:
as agreed, bumped minor version instead of major

salieri11 commented:
jenkins please retest this

salieri11 commented:
jenkins retest this please

salieri11 commented:
v2 related commits squashed

vshankar commented on src/test/client/zerocopy_sanity.cc:
```diff
@@ -20,20 +20,27 @@
+  io_info.base_info.iovcnt = 0;
+  io_info.base_info.write = false;
   io_info.zerocopy = true;
+  res = ceph_ll_readv_writev_v2(cmount, &io_info.base_info, sizeof(ceph_ll_io_info_v2));
```
How is this returning -EINVAL? info_size is sizeof(v2) struct and  the check returns -EINVAL if

> if (!info || info_size < sizeof(ceph_ll_io_info)) {

salieri11 replied:
Ah, this line asserts the zerocopy params that are validates in the client:
https://github.com/ceph/ceph/blob/458a032e2e1accb3119d3d59bb7de31cdf99f8fd/src/client/Client.cc#L11520
The code you mentioned is asserted here
https://github.com/ceph/ceph/blob/458a032e2e1accb3119d3d59bb7de31cdf99f8fd/src/test/client/zerocopy_sanity.cc#L41
where the size is set to be less then required minimum

vshankar commented on src/include/cephfs/libcephfs.h:
```diff
@@ -119,19 +119,23 @@ struct ceph_snapdiff_entry_t {
+	bool syncdataonly;
+};
+
+struct ceph_ll_io_info_v2 {
```
I guess this has to be done this way since this is a C-compat header (else, subclassing would have been a neat approach), yes?

salieri11 replied:
true, C header implies this way of API versioning

salieri11 commented:
Rebase on main

salieri11 committed:
libcephfs: dont't call callback for ZC blocking IO

Fixes: https://tracker.ceph.com/issues/69919
Signed-off-by: Igor Golikov <igolikov@ibm.com>

vshankar commented:
> @ffilz any news on the testing?

Update: I am finishing up with adding the relevant test to be able to start testing IOs on Ganesha clients with async+zeropcopy combination.

vshankar commented:
I am working on adding tests to run workloads with nfs-ganesha clients with async/zerocopy configs. See tracker: https://tracker.ceph.com/issues/73172

batrick reviewed (changes requested):
Development should halt on this PR until we get clarity around some issues:

- Any new API for zerocopy should not require turning off `client_oc`.
- I think `zerocopy` should be renamed `direct`. Should this not just be communicated via the `O_DIRECT` flag on `ll_open`? The current behavior is that libcephfs will treat O_DIRECT as a synchronous read/write EXCEPT the new async interfaces (written for ganesha) which implicitly copy the buffer for write. I'm thinking we should unify around the common terminology and file system interface: O_DIRECT.
- My understanding is that Ganesha will not respond to NFS Clients with unstable writes with the current use of the async interface. (Due to turning the object cacher off.) This means that Ganesha is effectively getting synchronous I/O with the OSDs (whoopsie!). This needs fixed by NFS-Ganesha. (Why can't it respond with UNSTABLE after libcephfs async write returns?)

To my frustration, this incremental creep of hand-wavey API change requests lacks any comprehensive clarity about what NFS-Ganesha requires. There must be a design document (in the tracker) that indicates (a) what's lacking in the API; (b) how other file systems (FSAL) solve it; (c) what changes are required. From what I understand, it's not just "zerocopy" that NFS-Ganesha wants but also (async read/write) callbacks to indicate: (a) the write is queued (i.e. we've issued the write to the `Filer`); (b) the write is durable. @ffilz / @kalebskeithley mentioned also a callback to indicate the buffer is released when the write is copied into the `oc`. However, that would happen already when `client_oc` is enabled (and `direct=false`)? I'm also not seeing a use-case for "unsafe/unstable callback" when that's adequately communicated already by "return from async write".

Note to @salieri11: please remove all references to downstream issues (IBM JIRA). Everything should be tracked in the upstream redmine issue.
